src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -14,35 +14,288 @@ namespace Microsoft.CodeAnalysis
+                    // and we'll remove any data from the cache.
+                    if (newTransitiveReferences != null && existingTransitiveReferences != null)
+                    {
+                        builder[projectIdToUpdate] = existingTransitiveReferences.Union(newTransitiveReferences);
+                    }
+                    else
+                    {
+                        // Either we don't know the full set of the new references being added, or don't know the existing set projectIdToUpdate.
+                        // In this case, just remove it
+                        builder.Remove(projectIdToUpdate);
+                    }
+                }
+            }
+
+            return builder.ToImmutable();
+        }
+
+        /// <summary>
+        /// Computes a new <see cref="_reverseTransitiveReferencesMap"/> for the addition of new projects.
+        /// </summary>
+        private static ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewReverseTransitiveReferencesMapForAdditionalProjectReferences(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingReverseTransitiveReferencesMap,
+            ProjectId projectId,
+            IReadOnlyList<ProjectId> referencedProjectIds)
+        {
+            // To update the reverse transitive map, we need to add the existing reverse transitive references of projectId to any of referencedProjectIds,
+            // and anything else with a reverse dependency on them. If we don't already know our reverse transitive references, then we'll have to instead remove
+            // the cache entries instead of update them. We'll fetch this from the map, and use "null" to indicate the "we don't know and should remove the cache entry"
+            // instead
+            existingReverseTransitiveReferencesMap.TryGetValue(projectId, out var newReverseTranstiveReferences);
+
+            if (newReverseTranstiveReferences != null)
+            {
+                newReverseTranstiveReferences = newReverseTranstiveReferences.Add(projectId);
+            }
+
+            // We'll now loop through each entry in our existing cache and compute updates. We'll accumulate them into this builder.
+            var builder = existingReverseTransitiveReferencesMap.ToBuilder();
+
+            foreach (var projectIdToUpdate in existingReverseTransitiveReferencesMap.Keys)
+            {
+                existingReverseTransitiveReferencesMap.TryGetValue(projectIdToUpdate, out var existingReverseTransitiveReferences);
+
+                // The projects who need to have their caches updated are projectIdToUpdate (since we're obviously updating it!)
+                // and also anything that depended on us.
+                if (referencedProjectIds.Contains(projectIdToUpdate) || existingReverseTransitiveReferences?.Overlaps(referencedProjectIds) == true)
```

Comment:
i don't understand why this is 'overlaps' here... but not contains.  It feels like this algorithm be symmetric to the forward-direction algorithm above.